clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+        {
+            return chunk_instr_count_;
+        }
+        uint64_t
+        get_page_size() const override
+        {
+            return page_size_;
+        }
+
+        // Supplied for better error messages from the client.
+        std::string
+        get_source_file() const
+        {
+            return cur_src_path_;
+        }
+
+    protected:
+        scheduler_tmpl_t<RecordType, ReaderType> *scheduler_ = nullptr;
+        int ordinal_ = -1;
+        int verbosity_ = 0;
+        uint64_t cur_ref_count_ = 0;
+        uint64_t cur_instr_count_ = 0;
+        uint64_t last_timestamp_ = 0;
+        std::string cur_src_path_;
+        // Remember top-level headers for the memtrace_stream_t interface.
+        uint64_t version_ = 0;
+        uint64_t filetype_ = 0;
+        uint64_t cache_line_size_ = 0;
+        uint64_t chunk_instr_count_ = 0;
+        uint64_t page_size_ = 0;
+    };
+
+    scheduler_tmpl_t()
+    {
+    }
+    scheduler_tmpl_t(int verbosity)
+        : verbosity_(verbosity)
+    {
+    }
+    virtual ~scheduler_tmpl_t()
+    {
+    }
+
+    virtual scheduler_status_t
+    init(std::vector<input_workload_t> &workload_inputs, int output_count,
+         scheduler_options_t options);
+
+    virtual stream_t *
+    get_stream(int ordinal)
+    {
+        if (ordinal < 0 || ordinal >= static_cast<int>(outputs_.size()))
+            return nullptr;
+        return &outputs_[ordinal].stream;
+    }
+
+    std::string
+    get_error_string()
+    {
+        return error_string_;
+    }
+
+protected:
+    typedef scheduler_tmpl_t<RecordType, ReaderType> sched_type_t;
+
+    struct input_info_t {
+        std::unique_ptr<ReaderType> reader;
+        std::unique_ptr<ReaderType> reader_end;
+        memref_tid_t tid;
+        // If non-empty these records should be returned before incrementing
+        // the reader.
+        std::queue<RecordType> queue;
+        std::set<int> binding;
+        int priority = 0;
+        std::vector<range_t> regions_of_interest;
+        int cur_region = 0;
+        bool needs_init = false;
+        bool at_eof = false;
+    };
+
+    struct output_info_t {
+        output_info_t(scheduler_tmpl_t<RecordType, ReaderType> *scheduler, int ordinal,
+                      int verbosity = 0)
+            : stream(scheduler, ordinal, verbosity)
+        {
+        }
+        stream_t stream;
+        memref_tid_t cur_input_tid = INVALID_THREAD_ID;
+        // For static schedules we can populate this up front and avoid needing a
+        // lock for dynamically finding the next input, keeping things parallel.
+        std::vector<memref_tid_t> tids_;
+        ssize_t cur_tids_index_ = 0;
+    };
+
+    scheduler_status_t
+    open_readers(const std::string &path, std::vector<memref_tid_t> &workload_tids);
+
+    scheduler_status_t
+    open_reader(const std::string &path, std::vector<memref_tid_t> &workload_tids);
+
+    std::unique_ptr<ReaderType>
+    get_default_reader();
```

Comment:
I see these are not public functions, but it'll be good to add a short comment on all of these.

Reply:
Done.